xbmc/storage/windows/Win32StorageProvider.cpp

```diff
@@ -56,46 +53,119 @@ void CWin32StorageProvider::Initialize()
   if (SUCCEEDED(SHGetFolderPath(nullptr, CSIDL_PROFILE, nullptr, 0, profilePath)) ||
       GetEnvironmentVariable(L"USERPROFILE", profilePath, MAX_PATH) > 0)
     share.strPath = KODI::PLATFORM::WINDOWS::FromW(profilePath);
   else
-#endif
     share.strPath = CSpecialProtocol::TranslatePath("special://home");
   share.strName = g_localizeStrings.Get(21440);
   share.m_ignore = true;
   share.m_iDriveType = CMediaSource::SOURCE_TYPE_LOCAL;
   localDrives.push_back(share);
 
-#ifndef TARGET_WINDOWS_STORE
-  CWIN32Util::GetDrivesByType(localDrives, LOCAL_DRIVES);
-#endif
+  GetDrivesByType(localDrives, LOCAL_DRIVES);
 }
 
 void CWin32StorageProvider::GetRemovableDrives(VECSOURCES &removableDrives)
 {
-  CWIN32Util::GetDrivesByType(removableDrives, REMOVABLE_DRIVES, true);
+  GetDrivesByType(removableDrives, REMOVABLE_DRIVES, true);
 }
 
 std::string CWin32StorageProvider::GetFirstOpticalDeviceFileName()
 {
-  return CWIN32Util::GetFirstOpticalDrive();
+  VECSOURCES vShare;
+  std::string strdevice = "\\\\.\\";
+  GetDrivesByType(vShare, DVD_DRIVES);
+
+  if(!vShare.empty())
+    return strdevice.append(vShare.front().strPath);
+  else
+    return "";
 }
 
 bool CWin32StorageProvider::Eject(const std::string& mountpath)
 {
-  if (!mountpath.empty())
+  using KODI::PLATFORM::WINDOWS::ToW;
+  if (mountpath.empty())
+  {
+    return false;
+  }
+
+  if( !mountpath[0] )
+    return false;
+
+  auto strVolFormat = ToW(StringUtils::Format("\\\\.\\%c:", mountpath[0]));
+
+  long DiskNumber = -1;
+
+  HANDLE hVolume = CreateFile(strVolFormat.c_str(), 0, FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, OPEN_EXISTING, NULL, NULL);
+  if (hVolume == INVALID_HANDLE_VALUE)
+    return false;
+
+  STORAGE_DEVICE_NUMBER sdn;
+  DWORD dwBytesReturned = 0;
+  long res = DeviceIoControl(hVolume, IOCTL_STORAGE_GET_DEVICE_NUMBER,NULL, 0, &sdn, sizeof(sdn), &dwBytesReturned, NULL);
+  CloseHandle(hVolume);
+  if ( res )
+    DiskNumber = sdn.DeviceNumber;
+  else
+    return false;
+
+  DEVINST DevInst = GetDrivesDevInstByDiskNumber(DiskNumber);
+
+  if ( DevInst == 0 )
+    return false;
+
+  ULONG Status = 0;
+  ULONG ProblemNumber = 0;
+  PNP_VETO_TYPE VetoType = PNP_VetoTypeUnknown;
+  wchar_t VetoName[MAX_PATH];
+  bool bSuccess = false;
+
+  CM_Get_Parent(&DevInst, DevInst, 0); // disk's parent, e.g. the USB bridge, the SATA controller....
+  CM_Get_DevNode_Status(&Status, &ProblemNumber, DevInst, 0);
+
+  for(int i=0;i<3;i++)
   {
-    return CWIN32Util::EjectDrive(mountpath[0]);
+    res = CM_Request_Device_Eject(DevInst, &VetoType, VetoName, MAX_PATH, 0);
+    bSuccess = (res==CR_SUCCESS && VetoType==PNP_VetoTypeUnknown);
+   if ( bSuccess )
+    break;
   }
-  return false;
+
+  return bSuccess;
 }
 
 std::vector<std::string > CWin32StorageProvider::GetDiskUsage()
 {
-  return CWIN32Util::GetDiskUsage();
+  std::vector<std::string> result;
+  ULARGE_INTEGER ULTotal = { { 0 } };
+  ULARGE_INTEGER ULTotalFree = { { 0 } };
+
+  std::unique_ptr<wchar_t> pcBuffer;
```

Comment:
`std::unique_ptr<wchar_t[]>` or it'll use the wrong deleter

Reply:
i'll fix this but it was in original code see https://github.com/xbmc/xbmc/pull/13254/files#diff-504bcfd2abeae7af84f0ab527d81e08fL396